### ISSUE TITLE
Graph Scope

### ISSUE THREAD

**Issue by tkohout** [hmlongco/Factory#42]
First of all, I would like to thank for this library. It's exactly what we were looking for.

I was wondering why Factory does not have built-in GraphScope. I know that Resolver had that kind of functionality so I am wondering about reasons why it's not included here. I think it's possible implement it in this way:

```
func mapInGraph<T>(_ resolve: (GraphIdentifier) -> T) -> T {
            // Will create new graph id if it is the root
            // Otherwise it will use graphID from parent graph
            let isRoot: Bool = Self.graphID == nil
            let graphID = Self.graphID ?? UUID()

            if isRoot {
                Self.graphID = graphID
            }

            // Resolve instance
            let instance = resolve(graphID)

            if isRoot {
                // reset graph ID when exiting the graph
                Self.graphID = nil
            }
            return instance
        }
```

cache would be then per graph:
```
private var cache: [GraphIdentifier: [Identifier: any Box]] = .init(minimumCapacity: 64)
```

and then resolve function has to be changed:

```
func resolve<T>(id: Identifier, factory: () -> T) -> T {
            defer { lock.unlock() }
            lock.lock()

            return mapInGraph { graphID in
                /// retrieve from cache based on graphID
                /// ...

                /// if not find in cache create new instance and store in cache based on grapID
                /// ...
            }
        }
```

There's some trickery that needs to be implemented because of multithreading but this is the basic concept. Do you see any problem with it?

**Comment by hmlongco** [comment]
It's not quite that simple. In Resolver graphs are implemented as another scope type, which means that we're only caching values that aren't otherwise cached. I mean, if it's a singleton, why recache the cached value?

In practice I found the depth of most iOS resolution trees were never that deep, and, as mentioned, in a lot of the cases where graphs could be useful the entities were already cached as singletons or shared values. Implementing them in Factory would also mean I'd need to resurrect the idea of the defaultScope and unique scopes for when you DON'T want an item cached.

Bottom line is that I didn't think it was worth the effort, performance hit, and added complexity.

**Comment by tkohout** [comment]
Thanks for detailed response. 

In my case I need to have 3 tabs which all contain the same screen with slightly different input parameters. Each of these screens have some objects that should be shared between classes in the stack but not between multiple screens. Previously we used hand written factories but it is A LOT of boilerplate.

So I am looking into how best create 3 separate stacks of the same objects with DI. And unless I am missing something I am not able to achieve this without the graph scope. 

I do have working implementation of graph scope, so this is not a feature request, just a question. You can close the issue unless there's something to add.

**Comment by hmlongco** [comment]
I played with an implementation and I'm not sure yet that I want to fold it in as the scope doesn't work as one might expect under all circumstances. Still under consideration.

**Comment by hmlongco** [comment]
Switch to the "graph" branch, pull it, and let me know what you think. Make sure you read the graph scope section added to the README.

**Comment by hmlongco** [comment]
Graph now available as option in 1.3.0.

**Comment by tkohout** [comment]
Thanks a lot. I've checked the code and it looks good.
